File: src/main/java/io/holunda/history/AsyncCamundaHistoryConfiguration.java
package io.holunda.history;

import org.camunda.bpm.engine.impl.history.handler.DbHistoryEventHandler;
import org.camunda.bpm.engine.spring.SpringProcessEngineConfiguration;
import org.camunda.bpm.spring.boot.starter.configuration.CamundaHistoryConfiguration;
import org.camunda.bpm.spring.boot.starter.configuration.impl.AbstractCamundaConfiguration;
import org.springframework.context.ApplicationEventPublisher;

public class AsyncCamundaHistoryConfiguration extends AbstractCamundaConfiguration implements CamundaHistoryConfiguration {

    private final ApplicationEventPublisher publisher;

    public AsyncCamundaHistoryConfiguration(ApplicationEventPublisher publisher) {
        this.publisher = publisher;
    }

    @Override
    public void preInit(SpringProcessEngineConfiguration configuration) {
        if (camundaBpmProperties.getHistoryLevel() != null) {
            configuration.setHistory(camundaBpmProperties.getHistoryLevel());
        }
        configuration.setHistoryEventHandler(new AsyncHistoryEventHandler(publisher, new DbHistoryEventHandler()));
    }
}

File: src/main/kotlin/io/holunda/history/AsyncCamundaHistoryConfiguration.kt
package io.holunda.history

import org.camunda.bpm.engine.ProcessEngine
import org.camunda.bpm.engine.impl.cfg.ProcessEngineConfigurationImpl
import org.camunda.bpm.spring.boot.starter.configuration.CamundaHistoryConfiguration
import org.camunda.bpm.spring.boot.starter.configuration.impl.AbstractCamundaConfiguration
import org.springframework.context.ApplicationEventPublisher

class AsyncCamundaHistoryConfiguration(
        private val publisher: ApplicationEventPublisher
) : CamundaHistoryConfiguration, AbstractCamundaConfiguration() {

    override fun preInit(configuration: ProcessEngineConfigurationImpl) {
        if (camundaBpmProperties.historyLevel != null) {
            configuration.history = camundaBpmProperties.historyLevel
        }
        configuration.historyEventHandler = AsyncHistoryEventHandler(publisher)
    }

    override fun postInit(processEngineConfiguration: ProcessEngineConfigurationImpl?) {
        super<AbstractCamundaConfiguration>.postInit(processEngineConfiguration)
    }

    override fun postProcessEngineBuild(processEngine: ProcessEngine?) {
        super<AbstractCamundaConfiguration>.postProcessEngineBuild(processEngine)
    }

}
